pandas/core/ops/common.py

```diff
@@ -0,0 +1,66 @@
+"""
+Boilerplate functions used in defining binary operations.
+"""
+from functools import wraps
+
+from pandas._libs.lib import item_from_zerodim
+
+from pandas.core.dtypes.generic import ABCDataFrame, ABCIndexClass, ABCSeries
+
+
+def unpack_and_defer(name: str):
+    """
+    Boilerplate for pandas conventions in arithmetic and comparison methods.
+
+    Parameters
+    ----------
+    name : str
+
+    Returns
+    -------
+    decorator
+    """
+
+    def wrapper(method):
+        return _unpack_and_defer(method, name)
+
+    return wrapper
+
+
+def _unpack_and_defer(method, name: str):
+    """
+    Boilerplate for pandas conventions in arithmetic and comparison methods.
+
+    Ensure method returns NotImplemented when operating against "senior"
+    classes.  Ensure zero-dimensional ndarrays are always unpacked.
+
+    Parameters
+    ----------
+    method : binary method
+    name : str
+
+    Returns
+    -------
+    method
+    """
+
+    is_cmp = name.strip("__") in {"eq", "ne", "lt", "le", "gt", "ge"}
+
+    @wraps(method)
+    def new_method(self, other):
+
+        if is_cmp and isinstance(self, ABCIndexClass) and isinstance(other, ABCSeries):
+            # For comparison ops, Index does *not* defer to Series
```

Comment:
Code clarity question: what are the types on `self`, and `other`?

I'm finding this a bit harder to follow than the rule "arrays defer to pandas objects, which will do the unpacking".

Would it be clearer to have a `unpack_and_defer_array` that's just for arrays, and a separate decorator that is for Series and Index (the Index one has to be for Series & Index, right? Since ops are shared between those two?)

Reply:
`self` can be any of our array-like classes: DataFrame, Series, FooIndex, BarExtensionArray.  `other` can be any.  I'll try to clarify this.

This is in my "would be nice, but isn't a blocker pile"